BUG: Fix for np.dtype(ctypes.Structure) does not respect _pack_ field

bmakos opened the pull request:
See #10532

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html#writing-the-commit-message
-->

bmakos committed:
BUG: Fix for np.dtype(ctypes.Structure) does not respect _pack_ field

See #10532

eric-wieser commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -33,17 +33,28 @@ def _from_ctypes_array(t):
-            raise TypeError(
+    # TODO: gh-10533
+    if hasattr(t,"_pack_"):
+        fields = {}
```
The `dtype({name: type})` constructor isn't safe to use, as it fails if one of the fields is called "formats".

I'd recommend using:
```python
np.dtype(dict(
    formats=[...],  # dtypes
    offsets=[...],  # integers
    names=[...],    # strings
    itemsize=ctypes.sizeof(t)
))
```
Which doesn't suffer from this fault.

bmakos replied:
Ok, I changed the code. I did not use ctypes.sizeof(t) as it does not take into consideration the _pack_ attribute it has (at least in my tests). 
For example: 
`class Foo(ctypes.Structure):
    _fields_ = [('one', ctypes.c_uint8), ('two', ctypes.c_uint32)]
    _pack_ = 6`

results in a Structure that has size 8 (while I expect size 12 - as it has 2 items in 2 6 long regions). I am a bit unsure if correct size is 12(items_nr * pack) or 10 (last offset + last item size). Or is the correct behavior to accept only `pack <= max(type size)` and use default pack when the `pack > max(type size)`?
This is how I understood the alignment for this case (2 items, sizes 1 and 4) is that the first item starts on 0, the second one starts on 6. To put it short, the items are placed at offsets divisible by pack. This is not happening for the above example, two has offset 4 in the ctypes structure (as PyCharm debugger shows it).

eric-wieser replied:
I think this line reveals the secret:

https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b/Modules/_ctypes/cfield.c#L154-L155

So yes, pack is only used if it's smaller than the default alignment.

bmakos replied:
Note: maybe I could add another (or several) test that would better force out errors. Something like a more complex structure: 

```
Class Foo(ctypes.Structure):
    _fields_ = [('1one', ctypes.c_uint8), ('2two', ctypes.c_uint16), ('3three', ctypes.c_uint8),('4four', ctypes.c_uint16),('5five', ctypes.c_uint32), ('6six', ctypes.c_uint32)]
    _pack_ = 3 
```

I am not sure if we want several tests for similar functionalities. Please let me know.

eric-wieser commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -33,17 +33,28 @@ def _from_ctypes_array(t):
-        fname, ftyp = item
-        fields.append((fname, dtype_from_ctypes_type(ftyp)))
+            fname, ftyp = item
+            fields[fname] = dtype_from_ctypes_type(ftyp), t._pack_
```
This looks wrong to me - you need to be keeping track of the size so far, and then rounding up to the next multiple of `_pack_` at each step.

bmakos replied:
I am not sure I understood the formula you provided for offset. This is how I would calculate it: 
`offset = (((offset + _pack_ - 1) // _pack_) + 1 ) * _pack_` - the first offset is set to 0 and this value is calculated after an item is added (as I need it's size).

eric-wieser replied:
Sorry, I meant that as two steps:
```
offset += field.itemsize
offset = (offset + _pack_ - 1) // _pack_ * _pack_)
```

eric-wieser commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -33,17 +33,28 @@ def _from_ctypes_array(t):
+        for item in t._fields_:
+            if len(item) > 2:
+                raise TypeError(
+                "ctypes bitfields have no dtype equivalent")
```
It might make sense to perform this checking loop separately, at the start of the function

bmakos replied:
Done, will be in next commit.

eric-wieser reviewed (changes requested):
Thanks for taking this on. This doesn't look quite right to me - and also, you need to go add / enable the test for it.

bmakos committed:
BUG: Fix for np.dtype(ctypes.Structure) does not respect _pack_ field

See #10532
Made several changes to offset and size calulation. First pull request
was far from correct.

eric-wieser commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -33,17 +33,43 @@ def _from_ctypes_array(t):
+        if hasattr(t,"_pack_"):
+            formats_array = []
+            offsets_array = []
+            names_array = []
```
I'd drop the `_array` part of the names here - it's confusing, because they're actually lists - and the `s` implies that it's a list anyway

bmakos replied:
Done. Also added an additional test.

eric-wieser commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -33,17 +33,43 @@ def _from_ctypes_array(t):
+                formats = formats_array,
+                offsets = offsets_array,
+                names = names_array,
+                itemsize = ctypes.sizeof(t)
```
PEP8 says no spaces around these eequals

bmakos replied:
Fixed.

eric-wieser commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -33,17 +33,43 @@ def _from_ctypes_array(t):
                 "ctypes bitfields have no dtype equivalent")
-        fname, ftyp = item
-        fields.append((fname, dtype_from_ctypes_type(ftyp)))
+    else:
```
No point in this else - you can remove it and unindent everything below

bmakos replied:
Indeed. Fixed.

eric-wieser commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -33,17 +33,43 @@ def _from_ctypes_array(t):
+            current_offset = 0
+            current_itemsize = 0
+            for item in t._fields_:
+                fname, ftyp = item
```
May as well combine this with the line above, as `for fname, ftype in t._fields_:`

bmakos replied:
Done.

eric-wieser commented on numpy/core/_dtype_ctypes.py:
```diff
@@ -33,17 +33,43 @@ def _from_ctypes_array(t):
-        fname, ftyp = item
-        fields.append((fname, dtype_from_ctypes_type(ftyp)))
+    else:
+        if hasattr(t,"_pack_"):
```
nit: space after the comma

bmakos replied:
Done.

eric-wieser reviewed (commented):
This looks great, thanks. Only stylistic remarks above.

If you think an extra test would be useful, go for it!

bmakos committed:
BUG: Fix for np.dtype(ctypes.Structure) does not respect _pack_ field

See #10532
Cosmetic changes + added an additional test with a more complex
structure.

eric-wieser reviewed (approved):
Looks great, thanks! Will leave this for a few days for others to look at.


Can you maybe update the release notes? <s>There should already be a section about improved ctypes compatibility</s> Nevermind, I put that off for #12272. I can take care of the notes for your PR at the same time.

bmakos commented:
@eric-wieser 
Thank you Eric for the help! I had a lot of fun working on this.

charris committed:
STY: Minor style fixups.

charris committed:
STY: Minor style fixups.

charris commented:
Thanks @bmakos, @eric-wieser .
